### PR TITLE
Update liblouis to version 3.8, add debug logging functionality and cleanup old workarounds that are no longer necessary

### LINKED ISSUE #4435: create grade 2 table for Arabic
**Reported by mohammed on 2014-09-02 07:28**
can anyone here get me started on creating a grade 2 Arabic table? I know the rules for that locale, but I couldn't understand the Liblouis documentation.

I can consider reasonable paid tutoring.

### LINKED ISSUE #4554: Send liblouis log output to the NVDA log
**Reported by jteh on 2014-10-16 23:18**
liblouis logs information which could be useful in debugging problems. We should send liblouis log output to the NVDA log at level debugWarning (or perhaps debug if it's too much). This is now possible with the logging callback introduced in liblouis 2.6.0.

### LINKED ISSUE #9013: upgrade Liblouis to version 3.8.0
Ho to all,
Today, the Liblouis 3.8.0 was released to the public.
It brings a ton of bubfixes, and a lot of renamed braille tables.
The biggest change to us is that pass1 only flags are removed completelly, and the braille tables where updated, and new tables were created.
Here is the excerpt from the changelog

The liblouis developer team is proud to announce the liblouis release 3.8.0. The release is available for download at:

    https://github.com/liblouis/liblouis/releases

Introduction
============

Liblouis is an open-source braille translator and back-translator. It features support for computer, literary and math braille, supports contracted and uncontracted translation for many, many languages[1].
It plays an important role in an open source accessibility stack and is used by screenreaders such as NVDA, Orca and JAWS. A companion project liblouisutdml[2] deals with formatting of braille.

Changes in this release
=======================

The major focus of this release is on braille table updates. There are major updates to German, Arabic, Chinese, Turkish, Dutch, Czech, Latvian, Spanish and Ethiopic. Some of these new tables have only been possible because Bert Frees fixed some nasty long standing bugs behind the scene. Also there is the usual assortment of code improvements and cleanups.

For a detailed list of all the changes refer to the list of closed issues[3].

New features
~~~~~~~~~~~~

- None

Bug fixes
~~~~~~~~~

- Fix support more than 4 classes thanks to Bert Frees.
- Fix capitalization of words that match `nocont' rules thanks to Bert
  Frees.

Braille table improvements
~~~~~~~~~~~~~~~~~~~~~~~~~~

- Defined the undefined character for the Czech tables thanks to Jan
  Hegr.
- Improvements to Unified English braille thanks to Mike Gray
- Updated the Dutch table to the new 2017.1 braille standard thanks to
  Davy Kager.
- Improvements to the Polish grade 1 table, to make back-translations
  of diacritics working
- Fixes to Latvian braille table thanks to Gatis Grintals and Artis
  Raugulis.
- Improvements to traditional Mainland Chinese braille and two-cell
  Chinese Braille thanks to Sunian Loomee.
- Update Bopomofo-based Chinese Braille to version 2018-11 thanks to
  Bo-Cheng Jhan
  - Correct the default braille pattern of many Chinese characters
  - Add various Chinese phrases involving exceptions of braille
    patterns
  - Modify dot patterns of dashes for readability reasons
- Added a table for Turkish grade 2 thanks to Uğur Gürbüz and Simon
  Aittamaa
- Major upgrade to the German tables. They have been upgraded to /Das
  System der deutschen Brailleschrift/ (2018). They are much smaller
  now as they are based on `lou_maketable'. As they now work for any
  locale (be it Switzerland or Germany) they have been merged into one
  set of tables for the different grades. Thanks to Christian
  Waldvogel.
- New table for Arabic contracted braille thanks to Ikrami Ahmad.
- New table for Arabic computer braille thanks to Ikrami Ahmad.
- Improvements to Ethiopic thanks to Tamru E. Belay.
- New table for Spanish contracted braille thanks to Juan Pablo Bello.

Other changes
~~~~~~~~~~~~~

- Updated the `lou_allround' and `lou_trace' test tools to include all
  the mode flags described in the documentation of the
  `lou_translateString()' function, thanks to Bue Vester-Andersen

Deprecation notice
~~~~~~~~~~~~~~~~~~

- None

Backwards incompatible changes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- The `pass1Only' flag has been deprecated for a while and is now
  removed from the code, thanks to Bue Vester-Andersen.

New, renamed or removed tables
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

New
---

  - tr-g2.ctb
  - ar-ar-g2.ctb
  - ar-ar-comp8.utb
  - es-g2.ctb

Renamed
-------

  - de-de-accents.cti -> de-accents.cti
  - de-de-g0.utb -> de-g0.utb
  - de-de-g1.ctb -> de-g1.ctb
  - de-de-g2.ctb -> de-g2.ctb
  - de-ch-g0.utb -> de-g0.utb
  - de-ch-g1.ctb -> de-g1.ctb
  - de-ch-g2.ctb -> de-g2.ctb

Removed
-------

  - ar-fa.utb
  - Es-Es-g1.utb

Next release
============

The next release will be published on March 4 2019 so please keep up the excellent work and keep those improvements coming.

Share and Enjoy!

-- Christian Egli, on behalf of the liblouis developers

Footnotes
=========

[1]  https://github.com/liblouis/liblouis/tree/master/tables
[2]  https://github.com/liblouis/liblouisutdml
[3]  https://github.com/liblouis/liblouis/milestone/18?closed=1

--
Christian Egli
Swiss Library for the Blind, Visually Impaired and Print Disabled Grubenstrasse 12, CH-8045 Zürich, Switzerland

For a description of the software, to download it and links to project pages go to http://liblouis.org
best,
Zvonimir

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Closes #9013 
Closes #4435
Closes #4554 

### Description of this pull request
1. Updates liblouis to version 3.8, thereby dealing with renamed tables. Also added grade 2 tables for Arabic and Spanish, and an Arabic 8 dot computer braille table.
2. Add the possibility to send liblouis' log output to the NVDA log. This is the reason why we are at liblouis commit 90a808bf, and not at 3.8.0 itself.
    The log output can be enabled by setting config.conf['debugLog']['louis'] to True. While @jcsteh suggested in https://github.com/nvaccess/nvda/issues/4554#issuecomment-315663267 to throw everything in at level debug, I think I've been able to come up with a sane mapping of louis levels to NVDA levels.
3. Now we are introducing more liblouis specific functionality into NVDA, I introduced a new louisHelper module. I added a translate method to that new module as well, to act as a convenience wrapper for louis.translate. The braille module now uses louisHelper.translate, and that function does the underlying magic to have NVDA work with the translation result in a correct way. I also investigated whether some workarounds needed to fix #2947 and #2466 were still necessary, and it looks like that they're fixed upstream. What also changed since then is that Liblouis now has a decent way of providing patches and getting assistance with bugs, thanks to github.
4. Something " I'd like to discuss and which is implemented in this pr as well, is signing liblouis dlls with the NV Access certificate. Now there are several liblouis dll builds in the wild, such as those from Microsoft, Freedom Scientific and even others, it might make sense to do this in order for NVDA specific liblouis builds to be recognized as such. Note that Microsoft also signs liblouis dlls that are included with Windows 10.

### Testing performed:
Unit tests, chose the new tables from the braille settings panel.

### Known issues with pull request:
* Once we agree that the fixes for #2947 and #2466 can be reverted, we should be alert for them returning
* I really wish we could have switched to UCS-4 builds of liblouis, however that's not yet possible and I doubt that that will ever make it into Python 2 builds of NVDA.

### Change log entry:
* New features
    + New braille tables: Arabic 8 dot computer braille, Arabic grade 2, Spanish grade 2. (#4435)

* Changes
    + Updated liblouis braille translator to version 3.8.0. (#9013)

* Changes for developers
    + You can now include log output from liblouis into the NVDA log by setting the louis boolean flag in the debugLogging section of the NVDA configuration. (#4554)

### CODE REVIEW

**Comment by ikrami1** [comment]
there is also some major improvements to the arabic grade1 table which are not mentioned in the release, including more compatibility with the unified Arabic braille code, fixing many problems regarding back-translation, and improving the overall readability of the table. i wish we see this Merged in NVDA as soon as possible>

**Comment by leonardder** [comment]
This won't be in NVDA 2018.4, but 2019.1 should work. Usually we don't mention fixes or improvements in liblouis in the NVDA what's new.

**Comment by dingpengyu** [comment]
@leonardder Thanks to leonardder's mood work liblouis3.8 This version of Chinese Braille also has a lot of modifications. I hope you can see him in the next alpha test build and test thank@

**Comment by leonardder** [comment]
I updated the initial description of this pr, as I've been able to expand it in a major way.
@dkager: since this pr involves some major changes, I'd like to have your review if possible. Don't feel forced to do it, though.

File: source/louisHelper.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+def _isDebug():
+	return config.conf["debugLog"]["louis"]
+
+def initialize():
+	# Register the liblouis logging callback.
+	louis.registerLogCallback(louis_log)
+	# Set the log level to all.
+	# The NVDA logging callback will filter messages appropriately,
+	# i.e. error messages will be logged at the error level.
+	louis.setLogLevel(louis.LOG_ALL)
```
Comment (michaelDCurran) [inline source/louisHelper.py]: Are you sure there is not a performance decrease with this? If there are a lot of log calls in Liblouis at level debug,  there will be many switches from c to Python, even if the log callback Python function returns early. I certainly agree we should enable for warning and error though.

Comment (leonardder) [reply source/louisHelper.py]: I haven't experienced any noticeable performance quirks. I guess we can do either of 4 things:

1. Leave as is
2. Initialize liblouis debug at louis level DEBUG, in which case the number of message drops significantly (i.e. as far as I know, all messages involved with a translation that succeeds, are logged below level debug and thus won't be logged in that case.
3. Do not initialize louis logging at all when the louis flag in the configuration is False. This requires a restart of NVDA for changes to the louis flag to take effect.
4. Revert the config spec change. Instead, initialize louis logging at level_none, in which case nothing is logged at all. However... that only adds the possibility to change the logging level at runtime.

Comment (michaelDCurran) [reply source/louisHelper.py]: Good to know you're not feeling any performance decrease. I'd still vote for going with option 2. Assuming there is no good reason why those messages below debug need to be easily accessed on any copy of NVDA.

**Review by michaelDCurran — APPROVED** [review]

**Comment by leonardder** [comment]
I don't think it's strictly necessary either, but better safe than sorry. There's nothing as scary as log output pumped to a callback function which python object died or something.

**Comment by DrSooom** [comment]
Is this PR already fully included into NVDA 2018.4? I ask because the changelog file wasn't updated.

**Comment by lukaszgo1** [comment]
It is included in 2019.1. It was just added to the wrong milestone.

**Pull request merged**